### PR TITLE
Add support for custom templates in 'rclone authorize'

### PR DESCRIPTION
**Pull request by hwittenborn**
#### What is the purpose of this change?

Adds the ability to provide a custom template to `rclone authorize` calls. Previously a hardcoded template was used, but a custom one may want to be used, such as for branding purposes when integrating Rclone into another app (such is my personal use case).

#### Was the change discussed in an issue or in the forum before?

See #6741 and https://forum.rclone.org/t/customize-html-page-used-in-responses-from-rclone-authorize/35128/1.

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

**Comment by hwittenborn** [comment]
@ncw: I'm just thinking about this now, but there's probably gonna need to be some stable API for this template, right? I'd also like to add some variables to it if you'd be fine with it, the main one I'm wanting at the moment is just the name of the remote type being authenticated. Any thoughts?

**Comment by hwittenborn** [comment]
On second thought I probably don't need nothing stabilized too much, I'll probably end up bundling rclone with my application so preserving the version shouldn't be an issue. But just having some extra data that I could use would be nice, I don't care too much about all the effort into making a stable interface at this moment though.

File: lib/oauthutil/oauthutil.go
Comment (albertony) [inline lib/oauthutil/oauthutil.go:29]: You should add a comment to this exported variable, which starts with the word "TemplatePath". The linter will complain, when I enable the workflow to run..

File: lib/oauthutil/oauthutil.go
Comment (albertony) [inline lib/oauthutil/oauthutil.go:601]: Error messages should start with lowercase letter (lint issue).

**Comment by ncw** [comment]
@hwittenborn we can document some template variables that will be available - that would be a great idea, and feel free to add more.

I'll allow the CI to run now so you can see the lint errors!

I'm not 100% happy with the way the template name gets passed in, I think there must be a better way, but I need to think about it :-)

**Comment by hwittenborn** [comment]
> we can document some template variables that will be available - that would be a great idea, and feel free to add more.

On that note the `AuthResult` struct is capitalized so it's part of rclone's public API, correct? I'm just asking because I'd like to rename some of the fields in the struct (like `Name` to `ErrMsg`) so they better fit their purpose.

Also - it looked like some of the fields in the struct (`HelpURL`, and `Err`) are never added to the struct in the `oathutil.go` file. Is that intentional, or are they just there from a past time when they were used or something? If they're never gonna be used, I'd rather just remove them as part of the public API being created.

> I'm not 100% happy with the way the template name gets passed in

Is it the way it's passed on the CLI, how it's passed between functions, or what's it you don't like about it?

File: cmd/authorize/authorize.go
Comment (albertony) [inline cmd/authorize/authorize.go:11]: The lint error is because the imports should be in alphabetical order, I'm using VSCode with go extension which does this automatically on save, but I think you can use the go fmt tool to do it file by file as well:

`gofmt -w cmd/authorize/authorize.go`

File: lib/oauthutil/oauthutil.go
Comment (albertony) [inline lib/oauthutil/oauthutil.go:16]: Same here.

**Comment by ncw** [comment]
> > I'm not 100% happy with the way the template name gets passed in
> 
> Is it the way it's passed on the CLI, how it's passed between functions, or what's it you don't like about it?

I think the name of the template file needs to be passed as a config parameter. We should also allow this to be passed in its entirety also which will make usage via the API easier.

So add some more definitions like this

```diff
--- a/fs/config/config.go
+++ b/fs/config/config.go
@@ -58,6 +58,12 @@ const (
 
 	// ConfigAuthNoBrowser indicates that we do not want to open browser
 	ConfigAuthNoBrowser = "config_auth_no_browser"
+
+	// ConfigTemplate template contents to be used in the config webserver
+	ConfigTemplate = "config_template"
+
+	// ConfigTemplateFile path to a template file to use for the config webserver
+	ConfigTemplateFile = "config_template_file"
 )
 
 // Storage defines an interface for loading and saving config to
```

And read them out of `m` in `lib/oauthutil`.

Passing the config via a command line parameter will mean it **only** works for `rclone authorize` and not for `rclone config` or use via the API.

You'd then do something like this in `rclone authorize`

```diff
--- a/cmd/authorize/authorize.go
+++ b/cmd/authorize/authorize.go
@@ -39,6 +39,6 @@ Use --template to generate HTML output via a custom Go template. If a blank stri
 	},
 	RunE: func(command *cobra.Command, args []string) error {
 		cmd.CheckArgs(1, 3, command, args)
-		return config.Authorize(context.Background(), args, noAutoBrowser)
+		return config.Authorize(context.Background(), args, noAutoBrowser, template)
 	},
 }
```

And something like this in `config.Authorize`

```diff
--- a/fs/config/authorize.go
+++ b/fs/config/authorize.go
@@ -15,7 +15,7 @@ import (
 //	rclone authorize "fs name"
 //	rclone authorize "fs name" "base64 encoded JSON blob"
 //	rclone authorize "fs name" "client id" "client secret"
-func Authorize(ctx context.Context, args []string, noAutoBrowser bool) error {
+func Authorize(ctx context.Context, args []string, noAutoBrowser bool, template string) error {
 	ctx = suppressConfirm(ctx)
 	ctx = fs.ConfigOAuthOnly(ctx)
 	switch len(args) {
@@ -40,6 +40,9 @@ func Authorize(ctx context.Context, args []string, noAutoBrowser bool) error {
 	if noAutoBrowser {
 		inM[ConfigAuthNoBrowser] = "true"
 	}
+	if template != "" {
+		inM[ConfigTemplateFile] = template
+	}
 
 	// Add extra parameters if supplied
 	if len(args) == 2 {
```

Yes it is rather convoluted but the rclone config system is very flexible!


> On that note the `AuthResult` struct is capitalized so it's part of rclone's public API, correct? I'm just asking because I'd like to rename some of the fields in the struct (like `Name` to `ErrMsg`) so they better fit their purpose.

Yes `Name` does look wrong now! I'd call it `Error`.

Rclone doesn't guarantee a stable public API so you can rename things if you want.
 
> Also - it looked like some of the fields in the struct (`HelpURL`, and `Err`) are never added to the struct in the `oathutil.go` file. Is that intentional, or are they just there from a past time when they were used or something? If they're never gonna be used, I'd rather just remove them as part of the public API being documented.

I think we can lose `Err` - we have a textual version in `Error`. I think I'd probably leave HelpURL just in case we do decide to add some URLs (would be nice!)

File: cmd/authorize/authorize.go
Comment (hwittenborn) [reply cmd/authorize/authorize.go:11]: I ran `go fmt` from the root of the repository, I'm not sure why it didn't catch it. I'll get that fixed though.

**Comment by hwittenborn** [comment]
I just pushed some changes, but I won't be able to test them until I'm home in about two hours. It compiled and all that just fine though, I just got to make sure it's all working fine.

**Comment by ncw** [comment]
@hwittenborn ping me here when you want me to give it another review.

**Comment by hwittenborn** [comment]
Sorry for the wait again @ncw, I got a bit sidetracked on my application that I'm using rclone with, but I just confirmed everything is working locally, so it should be ready for another review.

**Comment by hwittenborn** [comment]
@ncw: Any update on this?

**Review by ncw — APPROVED** [review]
I think this looks perfect now - thank you :-)

I will merge ~~now~~ after the CI passes.

**Comment by hwittenborn** [comment]
Sorry for the delay @ncw, this should be fixed now though (my googling showed it just being some `gofmt` errors, but I guess I'll find out once the CI runs :p).

**Comment by ncw** [comment]
Thank you for doing this and sorry it has taken a while to merge :-)

**Comment by hwittenborn** [comment]
All good @ncw :) Any idea when it'll land in a release though? I'd like to start including it with my application, but ideally I can pull rclone from a release first to do that.

**Comment by ncw** [comment]
1.62 is due in 10 days or so - it will land then :-) In the mean time it is in the [the latest beta](https://beta.rclone.org/).

**Pull request merged**
